src/puppetlabs/puppetdb/jdbc.clj

```diff
@@ -127,17 +127,43 @@
    #(and (map? %)
          (valid-jdbc-query? (:results-query %)))))
 
-(defn convert-result-arrays
-  "Converts Java and JDBC arrays in a result set using the provided function
-  (eg. vec, set). Values which aren't arrays are unchanged."
-  ([result-set]
-     (convert-result-arrays vec result-set))
-  ([f result-set]
-     (let [convert #(cond
-                     (kitchensink/array? %) (f %)
-                     (isa? (class %) java.sql.Array) (f (.getArray %))
-                     :else %)]
-       (map #(kitchensink/mapvals convert %) result-set))))
+(defn convert-if-array
+  ([x] (convert-if-array vec x))
+  ([convert x]
+   (cond
+     (kitchensink/array? x) (convert x)
+     (isa? (class x) java.sql.Array) (convert (.getArray x))
+     :else x)))
+
+(defn call-with-query-rows
```

Comment:
The other one streams too, I just thought this name was clearer, and it allows you to ask for arrays instead of maps if that's all you want/need.  And it's somewhat following the Scheme convention to scope operations on a resource with a lifetime (cf. `call-with-output-file`).  I figured the docstring might be sufficient, but if we wanted to change the name, I wondered about `call-with-lazy-query-rows`.